### ISSUE TITLE
Bug: Receiver app stream handlers fail when using keyPrefix in redis client options

### ISSUE THREAD

**Issue by zeejers** [tamimaj/nestjs-redis-streams#19]
### Summary
In our applications we always use keyPrefix in our Redis options. However, it seems like keyPrefix is incompatible with the implementation of nestjs-redis-stream's RedisStreamStrategy.

### Steps to Repeat

Configure a microservice
```typescript
  app.connectMicroservice<MicroserviceOptions>({
    strategy: new RedisStreamStrategy({
      connection: { host: "localhost", port: 6379, keyPrefix: "my-prefix"},
      streams: {
        block,
        consumer,
        consumerGroup,
        deleteMessagesAfterAck: false,
      },
    }),
  });
  await app.startAllMicroservices();
```

Configure a controller
```typescript
@Controller()
export class StreamCommandController {
  @RedisStreamHandler('my-prefix:stream:control')
  async handleCommand(@Payload() data: any, @Ctx() ctx: RedisStreamContext): Promise<any> {
    console.log({ message: 'Command received', data, ctx });
    return [];
  }
}
```

Results in this error on app boot and listening fails.
```
ERROR [2024-08-28 02:14:55]: NOGROUP No such key 'my-prefix:stream:control' or consumer group 'api-group' in XREADGROUP with GROUP option
    context: "Server"
    err: {
      "type": "ReplyError",
      "message": "NOGROUP No such key 'my-prefix:stream:control' or consumer group 'api-group' in XREADGROUP with GROUP option",
      "stack":
          ReplyError: NOGROUP No such key 'my-prefix:stream:control' or consumer group 'api-group' in XREADGROUP with GROUP option
      "command": {
        "name": "xreadgroup",
        "args": [
          "GROUP",
          "api-group",
          "api",
          "BLOCK",
          "5000",
          "STREAMS",
          "my-prefix:stream:control",
          ">"
        ]
      }
    }
```

If we change the microservice controller to NOT include the prefix in the RedisStreamHandler decorator param, like this:
```typescript
@Controller()
export class StreamCommandController {
  @RedisStreamHandler('stream:control')
  async handleCommand(@Payload() data: any, @Ctx() ctx: RedisStreamContext): Promise<any> {
    console.log({ message: 'Command received', data, ctx });
    return [];
  }
}
```
Then the listener init on boot works fine, but then, sending a command like this
`XADD my-prefix:stream:control * data '{"command":"say_hello"}'`

We get this error
```
ERROR [2024-08-28 02:19:30]: handler is not a function
    context: "Server"
    err: {
      "type": "TypeError",
      "message": "handler is not a function",
      "stack":
          TypeError: handler is not a function
              at RedisStreamStrategy.<anonymous>
              at Generator.next (<anonymous>)
              at fulfilled
              at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

No config seems to allow keyPrefix to work properly with this lib when using the receiver.

**Comment by tamimaj** [comment]
Hi @zeejers 
Thanks for reporting.

I have to dig further to see if ioredis lib supports the key prefix.

Also, try without supplying the key prefix in the config, and just  use it in the handler. 

If users:create works as handler, then my-prefix:stream:control should also works directly.

The lib will create consumer groups automatically on bootstrapping the server side for each handler if it does not exist already. You don't need to create manually.

Also please, take the example client app and send from the client side of the lib to your stream handler to test, and see if that works.

The lib has default serialization/deserialization logic, and the payloads by default needs a certain shape, unless you provide your own serialization and deserialization code. Check the example apps for the format.

Test these things and let me know!

Thank you very much.

**Comment by kenzik** [comment]
@tamimaj - ioredis appears to support "transparent" use of keyprefix:

```ts
// consume.ts
import Redis from 'ioredis';

const redis = new Redis({
  host: 'localhost',
  port: 6379,
  keyPrefix: 'myapp:',
});

const streamName = 'testStream';

async function consumeStream() {
  console.log('Starting to consume from stream...');

  let lastId = '0-0';

  while (true) {
    try {
      const result = await redis.xread('BLOCK', 5000, 'STREAMS', streamName, lastId);

      if (result) {
        const [stream, messages] = result[0];

        for (const [id, fields] of messages) {
          console.log(`Received message with ID ${id}:`);
          for (let i = 0; i < fields.length; i += 2) {
            console.log(`  ${fields[i]}: ${fields[i + 1]}`);
          }
          lastId = id;
        }
      }
    } catch (error) {
      console.error('Error reading from stream:', error);
    }
  }
}

consumeStream().catch(console.error);
```

Launch: `tsx consume.ts`

```ts
// publish.ts
import Redis from 'ioredis';

const redis = new Redis({
  host: 'localhost',
  port: 6379,
  keyPrefix: 'myapp:',
});

const streamName = 'testStream';

async function publishTestData() {
  console.log('Starting to publish test data...');

  let counter = 0;

  while (true) {
    const message = {
      timestamp: Date.now().toString(),
      data: `Test message ${counter}`,
    };

    try {
      const id = await redis.xadd(streamName, '*', 'timestamp', message.timestamp, 'data', message.data);
      console.log(`Published message with ID ${id}`);
      counter++;
    } catch (error) {
      console.error('Error publishing to stream:', error);
    }

    // Wait for 2 seconds before publishing the next message
    await new Promise((resolve) => setTimeout(resolve, 2000));
  }
}

publishTestData().catch(console.error);
```

Launch: `tsx publish.ts`

Then in Redis CLI, to confirm, this indeed functions as expected:

`XREAD BLOCK 0 STREAMS myapp:testStream $`

Returns as expected:

```
{
  "myapp:testStream": [
    {
      "id": "1724835819648-0",
      "data": {
        "timestamp": "1724835819647",
        "data": "Test message 4"
      }
    }
  ]
}
```

**Comment by tamimaj** [comment]
@kenzik @zeejers  I will debug the lib locally with the key prefix and see what is going on.

**Comment by tamimaj** [comment]
@kenzik @zeejers 

So guys, I could replicate the error locally. The problem is this, the key prefix is on ioredis but I am using @types/ioredis which did not has the types for keyprefix. 

I removed it locally and there is no problem in supplying key prefix (from types perspective)

Until here is ok, the problem is, the lib try to creates consumer groups for all the streams and groups we have, otherwise XreadGroup will fail like the error you provided. However, the lib is not coded to recognize if there is a a key prefix supplied then the stream name will not be "users:create" will be "prefix:users:create", so when we XReadGroup, there is nothing like this one since the lib did not create it before starting out.

My other problem is that the XreadGroup command is behaving weirdly when supplying key prefix, it does not add the colon automatically after the prefix, so when I am supplying "my-key-prefix" in the config, is trying to read "my-key-prefixusers:create", it does not supply the colon and I am not controlling that in the lib, this is done by ioredis when reading behind the scenes.

To get the key prefix working, we have to debug why the read command is not supplying the ":" after the prefix and trying to read without it. Then we have to change the code that creates the consumer groups to detect the prefix and create consumer groups for the streams names coming from handler, prefixed with the prefix.

Same logic should be applied to the client side of the lib, when supplying key prefix to the config, when sending or emitting, we have to append the prefix to the stream name, otherwise won't work.

I don't think I have the time now to work on that. I think you can add your prefixes manually in the handlers while listening and to your client side when sending or emitting messages.

It is not perfect but will do the same thing, as far as I am concerned. 
If you want to work on  PR for that, I can push the playground branch I was testing on, and you can start from there!

**Comment by kenzik** [comment]
Thanks, @tamimaj - can take a look at your working branch when available. We have a hack in place so we can move forward with our work, but it's not PR ready.

**Comment by tamimaj** [comment]
Here it is @kenzik 
https://github.com/tamimaj/nestjs-redis-streams/tree/bug-1-prefix

Removed the IO redis types so you can supply the prefix without type issues.
Try to run the example apps locally, if you don't have streams called "my-key-prefixusers:create" or "my-key-prefixusers:update" the error provided by @zeejers appears, which basically mean we are trying to read a stream with consumer group that does not exist. 

If we make the handler "my-key-prefixusers:create", then the ioredis tries to read from "my-key-prefixmy-key-prefixusers:create", so double the prefix :)

The rest is what I explained from above.

**Comment by zeejers** [comment]
@tamimaj I submitted a PR: https://github.com/tamimaj/nestjs-redis-streams/pull/20

Explanation is in the pr.

**Comment by tamimaj** [comment]
@zeejers Thanks man for the PR addition. It is merged and published now, 1.2.0.
